libcudacxx/include/cuda/std/__atomic/functions/host.h

```diff
@@ -0,0 +1,238 @@
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#include <cuda/std/__atomic/order.h>
+#include <cuda/std/__atomic/platform.h>
+#include <cuda/std/__type_traits/enable_if.h>
+#include <cuda/std/__type_traits/is_floating_point.h>
+#include <cuda/std/__type_traits/remove_cvref.h>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+_CCCL_DIAG_PUSH
+_CCCL_DIAG_SUPPRESS_CLANG("-Watomic-alignment")
+
+template <typename _Tp>
+struct __atomic_alignment_wrapper
+{
+  _CCCL_ALIGNAS(sizeof(_Tp)) _Tp __atom;
+};
+
+template <typename _Tp>
+__atomic_alignment_wrapper<__remove_cv_t<_Tp>>& __atomic_auto_align(_Tp* __a)
+{
+  using __aligned_t = __atomic_alignment_wrapper<__remove_cv_t<_Tp>>;
+  return *reinterpret_cast<__aligned_t*>(__a);
+};
+template <typename _Tp>
+const __atomic_alignment_wrapper<__remove_cv_t<_Tp>>& __atomic_auto_align(const _Tp* __a)
+{
+  using __aligned_t = const __atomic_alignment_wrapper<__remove_cv_t<_Tp>>;
+  return *reinterpret_cast<__aligned_t*>(__a);
+};
+template <typename _Tp>
+volatile __atomic_alignment_wrapper<__remove_cv_t<_Tp>>& __atomic_auto_align(volatile _Tp* __a)
+{
+  using __aligned_t = volatile __atomic_alignment_wrapper<__remove_cv_t<_Tp>>;
+  return *reinterpret_cast<__aligned_t*>(__a);
+};
+template <typename _Tp>
+const volatile __atomic_alignment_wrapper<__remove_cv_t<_Tp>>& __atomic_auto_align(const volatile _Tp* __a)
+{
+  using __aligned_t = const volatile __atomic_alignment_wrapper<__remove_cv_t<_Tp>>;
+  return *reinterpret_cast<__aligned_t*>(__a);
+};
+
+// Guard ifdef for lock free query in case it is assigned elsewhere (MSVC/CUDA)
+inline void __atomic_thread_fence_host(memory_order __order)
+{
+  __atomic_thread_fence(__atomic_order_to_int(__order));
+}
+
+inline void __atomic_signal_fence_host(memory_order __order)
+{
+  __atomic_signal_fence(__atomic_order_to_int(__order));
+}
+
+template <typename _Tp, typename _Up>
+inline void __atomic_store_host(_Tp* __a, _Up __val, memory_order __order)
+{
+  __atomic_store(
+    &__atomic_auto_align<_Tp>(__a), &__atomic_auto_align<__remove_cv_t<_Tp>>(&__val), __atomic_order_to_int(__order));
+}
+
+template <typename _Tp>
+inline auto __atomic_load_host(_Tp* __a, memory_order __order) -> __remove_cv_t<_Tp>
+{
+  __remove_cv_t<_Tp> __ret;
+  __atomic_load(
+    &__atomic_auto_align<_Tp>(__a), &__atomic_auto_align<__remove_cv_t<_Tp>>(&__ret), __atomic_order_to_int(__order));
+  return __ret;
+}
+
+template <typename _Tp, typename _Up>
+inline auto __atomic_exchange_host(_Tp* __a, _Up __val, memory_order __order) -> __remove_cv_t<_Tp>
+{
+  __remove_cv_t<_Tp> __ret;
+  __atomic_exchange(&__atomic_auto_align<_Tp>(__a),
+                    &__atomic_auto_align<__remove_cv_t<_Tp>>(&__val),
+                    &__atomic_auto_align<__remove_cv_t<_Tp>>(&__ret),
+                    __atomic_order_to_int(__order));
+  return __ret;
+}
```

Comment:
@miscco @griwes 
Time for excuses: `__atomic_auto_align` why did I write this?

Because alignment information is lost as we descend into the abyss. And if we continued to link we'd end up requiring libatomic. There could be a better way to write this. Any ideas?   

https://gcc.godbolt.org/z/vdxffdnKc